lib/fluent/plugin/out_datadog.rb

```diff
@@ -56,147 +72,303 @@ def multi_workers_ready?
     true
   end
 
-  def new_client
-    if @use_ssl
-      context    = OpenSSL::SSL::SSLContext.new
-      socket     = TCPSocket.new @host, @ssl_port
-      ssl_client = OpenSSL::SSL::SSLSocket.new socket, context
-      ssl_client.connect
-      return ssl_client
-    else
-      return TCPSocket.new @host, @port
-    end
+  def formatted_to_msgpack_binary?
+    true
   end
 
   def start
     super
-    @my_mutex = Mutex.new
-    @running = true
-
-    if @tcp_ping_rate > 0
-      @timer = Thread.new do
-        while @running do
-          messages = Array.new
-          messages.push("fp\n")
-          send_to_datadog(messages)
-          sleep(@tcp_ping_rate)
-        end
-      end
-    end
+    @client = new_client(log, @api_key, @use_http, @use_ssl, @no_ssl_validation, @host, @ssl_port, @port, @use_compression)
   end
 
   def shutdown
     super
-    @running = false
-    if @client
-      @client.close
-    end
+  end
+
+  def terminate
+    super
+    @client.close if @client
   end
 
   # This method is called when an event reaches Fluentd.
   def format(tag, time, record)
     # When Fluent::EventTime is msgpack'ed it gets converted to int with seconds
     # precision only. We explicitly convert it to floating point number, which
     # is compatible with Time.at below.
-    return [tag, time.to_f, record].to_msgpack
+    record = enrich_record(tag, time.to_f, record)
+    if @use_http
+      record = Yajl.dump(record)
+    else
+      if @use_json
+        record = "#{api_key} #{Yajl.dump(record)}"
+      else
+        record = "#{api_key} #{record}"
+      end
+    end
+    [tag, time.to_f, record].to_msgpack
```

Comment:
If you use only record elements, you can pack record only into msgpack format.
This omission might be helpful to reduce buffer memory/file space.

```ruby
[record].to_msgpack
```

Reply:
Correct, updated